### PR TITLE
Cosim update

### LINKED ISSUE #14: Conv2D COSIM kernel building failed
## 🐛 Bug

Kernel building failed when running on COSIM. Undefined references to `bsg_printf` and `std::ios_base::Init::~Init()`

## To Reproduce

1. Build PyTorch for COSIM
1. Goto `hammerblade/torch`
1. `make kernel.riscv`

## Possible Reasons

1. `<iostream>` is included in `kernel_conv.cpp`
1. `bsg_printf` is not included correctly in `Makefile`

### PR DESCRIPTION
**Pull request by vb000**
1. Conv2d running on COSIM.
2. Updated COSIM to use DUSTIN's cool changes that gives about ~3x speedup to the plain cosim.
3. Used compilation definitions in `bsg_manycore` to compile riscv kernels for COSIM to maintain single source of truth for what flags should be used when compiling for Manycore.

Apparently conv2d tests I added take too long on cosim... these have been running whole night and this is the status:
```
BSG INFO: test_python Regression Test (COSIMULATION)
BSG INFO: pytest_runner Regression Test

 files collected by pytest runner:

 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_abs.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_addmm.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_add_sub.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_conv2d.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_dot.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_fill_.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_hammerblade_backend.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_mm.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_mul_div.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_relu.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_resize.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_sigmoid.py
 /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim/hammerblade/torch/tests/test_view.py

 starting pytest ...

============================= test session starts ==============================
platform linux -- Python 3.6.8, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /mnt/users/ssd1/no_backup/bandhav/venv_pytorch_cosim/bin/python3
cachedir: .pytest_cache
rootdir: /mnt/users/ssd1/no_backup/bandhav/hb-pytorch-cosim
collecting ... collected 66 items

tests/test_abs.py::test_torch_abs_1 PASSED                               [  1%]
tests/test_abs.py::test_torch_abs_2 PASSED                               [  3%]
tests/test_addmm.py::test_torch_addmm_1 SKIPPED                          [  4%]
tests/test_addmm.py::test_torch_addmm_2 SKIPPED                          [  6%]
tests/test_add_sub.py::test_elementwise_add_1 PASSED                     [  7%]
tests/test_add_sub.py::test_elementwise_add_2 PASSED                     [  9%]
tests/test_add_sub.py::test_elementwise_add_3 PASSED                     [ 10%]
tests/test_add_sub.py::test_elementwise_add_4 PASSED                     [ 12%]
tests/test_add_sub.py::test_elementwise_in_place_add PASSED              [ 13%]
tests/test_add_sub.py::test_add_with_scalar PASSED                       [ 15%]
tests/test_add_sub.py::test_elementwise_sub_1 PASSED                     [ 16%]
tests/test_add_sub.py::test_elementwise_sub_2 PASSED                     [ 18%]
tests/test_add_sub.py::test_elementwise_sub_3 PASSED                     [ 19%]
tests/test_add_sub.py::test_elementwise_sub_4 PASSED                     [ 21%]
tests/test_add_sub.py::test_elementwise_in_place_sub PASSED              [ 22%]
tests/test_add_sub.py::test_sub_with_scalar PASSED                       [ 24%]
tests/test_conv2d.py::test_conv2d_1 PASSED                               [ 25%]
tests/test_conv2d.py::test_conv2d_2 PASSED                               [ 27%]
tests/test_conv2d.py::test_conv2d_3 PASSED                               [ 28%]
tests/test_conv2d.py::test_conv2d_4 PASSED                               [ 30%]
tests/test_conv2d.py::test_conv2d_batch_input_output PASSED              [ 31%]
tests/test_conv2d.py::test_conv2d_width_height_kernel 
```
`tests/test_conv2d.py::test_conv2d_batch_input_output` ran for hours before I slept and passed sometime in the night for reference.

### CODE REVIEW

File: hammerblade/torch/tests/test_conv2d.py [hunk 1]
```diff
@@ -43,13 +45,58 @@ def test_conv2d_4():
...
+    Large padding
+    """
+    kernel = torch.rand(2, 3, 3, 3)
+    inputs = torch.rand(2, 3, 5, 5)
+    padding = 3
+    stride = 3
+
+    _test_conv2d(inputs, kernel, padding, stride)
+
+@pytest.mark.skipif(os.environ.get('USE_HB_EMUL') is None, reason="Prohibitively slow on cosim")
```
Comment (vb000) [inline hammerblade/torch/tests/test_conv2d.py:99]: Added more small conv2d tests and excluded thorough tests when running on cosim, as they are prohibitively slow (~days). 

May be the later set is better suited for running on an F1 instance.

Comment (yodada) [reply hammerblade/torch/tests/test_conv2d.py:99]: yeah they are really slow XD

Mine has been running for more than 36hrs

**Comment by yodada** [comment]
Does it run on older version of bladerunner? like v3.6.1? since I think you mentioned adjustments on how we load the python script.

If it does not ... let me figure out why I cannot build bladerunner v4.0.0, and then I'll try merging this in. ;)

**Comment by vb000** [comment]
> Does it run on older version of bladerunner? like v3.6.1? since I think you mentioned adjustments on how we load the python script.
> 
v3.6.2 @yodada.

The only interface to COSIM could be `libbsg_manycore_runtime.so` and `test_loader.cosim`. And continuous effort is being made to make them efficient and fast. So, instead of having a parallel cosim build flow in hb-ptorch, I thought we could use `libbsg_manycore_runtime.so` and `test_loader.cosim` built in bsg_replicant. This pr does that.

File: hammerblade/torch/Makefile [hunk 2]
```diff
@@ -1,142 +1,56 @@
...
-################################################################################
-# Paths / Environment Configuration
-################################################################################
-_REPO_ROOT ?= $(shell git rev-parse --show-toplevel)
-CURRENT_PATH := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
+COSIM_EXE        := $(BRG_BSG_BLADERUNNER_DIR)/bsg_replicant/testbenches/python/test_loader
+BSG_MANYCORE_DIR := $(BRG_BSG_BLADERUNNER_DIR)/bsg_manycore
+KERNEL_DIR       := $(shell git rev-parse --show-toplevel)/hammerblade/torch/kernel
+SCRIPT_DIR       := $(shell git rev-parse --show-toplevel)/hammerblade/torch/
+SCRIPT           := pytest_runner
```
Comment (yodada) [inline hammerblade/torch/Makefile:7]: I think we need `IGNORE_CADENV           := 1` here. Otherwise it doesn't work for us.

**Review by yodada — APPROVED** [review]
Yeah we just need that `IGNORE_CADENV           := 1`. 

Otherwise it looks good to me!

File: hammerblade/torch/Makefile (hunk 2 above)
Comment (vb000) [reply hammerblade/torch/Makefile:7]: Added it. Is it possible to run one test on cosim to make sure it works for you? All cosim tests are passing for me.

File: hammerblade/torch/tests/test_conv2d.py (hunk 1 above)
Comment (yodada) [reply hammerblade/torch/tests/test_conv2d.py:99]: everything passed for me ;)

File: run-hb-pytest.sh [hunk 3]
```diff
@@ -9,4 +9,4 @@ while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symli
 done
 DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
 
-make -C $DIR/hammerblade/torch clean test_loader.cosim.log
+make -C $DIR/hammerblade/torch clean regression
```
Comment (yodada) [inline run-hb-pytest.sh:12]: Could you add `make -C $DIR/hammerblade/torch clean` to the end? I think after running this script, the work directory is no longer clean due to untracked files under hammerblade/torch

Comment (vb000) [reply run-hb-pytest.sh:12]: Sure, I always go inside `torch/` and run cosim followed by manual `make clean`. So, didn't notice this ;-).

Comment (yodada) [reply run-hb-pytest.sh:12]: appreciate it!

Comment (yodada) [reply run-hb-pytest.sh:12]: You can go ahead and merge once Travis finishes running ;)

Comment (vb000) [reply run-hb-pytest.sh:12]: Yes, of course.

**Pull request merged**